tools/upload.py

```diff
@@ -16,20 +17,67 @@
     sys.path.insert(0, toolspath + "/esptool") # Add esptool dir to search path
     import esptool # If this fails, we can't continue and will bomb below
 except:
-    sys.stderr.write("Error in command line, need pyserial path as 1st arg and esptool path as 2nd.\n")
+    sys.stderr.write("pyserial or esptool directories not found next to this upload.py tool.\n")
     sys.exit(1)
 
-fakeargs = [];
+cmdline = []
+write_option = ''
+erase_addr = ''
+erase_len = ''
+
 while len(sys.argv):
-    if sys.argv[0] == '--end':
-        esptool.main(fakeargs)
-        sys.argv.pop(0) # Remove --end
-        fakeargs = []
-    else:
-        # We silently replace the 921kbaud setting with 460k to enable backward
-        # compatibility with the old esptool-ck.exe.  Esptool.py doesn't seem
-        # work reliably at 921k, but is still significantly faster at 460kbaud.
-        thisarg = sys.argv.pop(0)
-        if thisarg == "921600":
-            thisarg = "460800"
-        fakeargs = fakeargs + [thisarg]
+    thisarg = sys.argv.pop(0)
+
+    # We silently replace the 921kbaud setting with 460k to enable backward
+    # compatibility with the old esptool-ck.exe.  Esptool.py doesn't seem
+    # work reliably at 921k, but is still significantly faster at 460kbaud.
+    if thisarg == "921600":
+        thisarg = "460800"
+
+    # 'erase_flash' command is translated to the write_flash --erase-all option
+    # https://github.com/esp8266/Arduino/issues/6755#issuecomment-553208688
+    if thisarg == "erase_flash":
+        write_option = '--erase-all'
+        thisarg = ''
+
+    if thisarg == 'erase_region':
+        erase_addr = sys.argv.pop(0)
+        erase_len = sys.argv.pop(0)
+        thisarg = ''
+
+    if thisarg == 'ignore_this_arg':
```

Comment:
I believe you can simply pass an empty value (0-length argv[] value) in boards.txt.py instead of 'ignore_this_arg' and be a little cleaner by removing this if block.  Your len(thisarg) will be false in this case and you'll be fine.  Or, depending on how Arduino pastes things together, the 0-len arg might actually be dropped completely when calling upload.py.